hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -1159,10 +1161,30 @@ protected boolean validateTimelineBeforeSchedulingCompaction(Option<String> inFl
    * @param writeStatuses {@code WriteStatus} from the write operation
    */
   private HoodieData<HoodieRecord> getRecordIndexUpdates(HoodieData<WriteStatus> writeStatuses) {
-    return writeStatuses.flatMap(writeStatus -> {
-      List<HoodieRecord> recordList = new LinkedList<>();
-      for (HoodieRecordDelegate recordDelegate : writeStatus.getWrittenRecordDelegates()) {
-        if (!writeStatus.isErrored(recordDelegate.getHoodieKey())) {
+    // 1. List<HoodieRecordDelegate>
+    // 2. Reduce by key: accept keys only when new location is not
+    return writeStatuses.map(writeStatus -> writeStatus.getWrittenRecordDelegates().stream().map(recordDelegate -> Pair.of(writeStatus, recordDelegate)))
+        .flatMapToPair(Stream::iterator)
```

Comment:
before calling reduce By key, shouldn't we do KeyBy record key for all recordDelegate and then call reduceByKey.

Reply:
Lets do something like this. might be simpler

HD<WriteStatus> 

HD<WS> map to HpairD<String(RK), HRD> 

reduceByKey(combine) -> record (only one record for one record key)

.values() should give us the final HD<HRD>